src/openrct2/peep/Guest.cpp

```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
     }
 
     CoordsXY targetLoc = { DestinationX & 0xFFE0, DestinationY & 0xFFE0 };
+    uint16_t stationBaseZ = ride->stations[0].GetBaseZ();
+    // var_37 is 3, 7, 11 or 15
+    uint8_t subTileIndex = Var37 / 4;
+    uint8_t openEdgesFlags = 0;
+    uint8_t openEdgesCount = 0;
 
-    int16_t stationBaseZ = ride->stations[0].GetBaseZ();
-
-    // Find the station track element
-    auto trackElement = map_get_track_element_at({ targetLoc, stationBaseZ });
-    if (trackElement == nullptr)
+    // Get info on open edges at the current subTile
     {
-        return;
-    }
+        TrackElement* trackElement = map_get_track_element_at({ targetLoc, stationBaseZ });
+        if (trackElement == nullptr)
+            return;
 
-    uint16_t mazeEntry = trackElement->GetMazeEntry();
-    // Var37 is 3, 7, 11 or 15
-    uint8_t hedges[4]{ 0xFF, 0xFF, 0xFF, 0xFF };
-    uint8_t openCount = 0;
-    uint8_t mazeReverseLastEdge = direction_reverse(MazeLastEdge);
-    for (uint8_t i = 0; i < 4; ++i)
-    {
-        if (!(mazeEntry & (1 << _MazeCurrentDirectionToOpenHedge[Var37 / 4][i])) && i != mazeReverseLastEdge)
+        uint16_t mazeEntry = trackElement->GetMazeEntry();
+        auto nearEntranceTile = ride->stations[0].Entrance.ToCoordsXYZD();
+        Direction entranceDirection = nearEntranceTile.direction;
+
+        // If the subTile is connected to the maze entrance, close the edge giving to it
+        nearEntranceTile += CoordsDirectionDelta[entranceDirection];
+        entranceDirection = direction_reverse(entranceDirection);
+        if (nearEntranceTile.x == targetLoc.x && nearEntranceTile.y == targetLoc.y
+            && (subTileIndex == entranceDirection || subTileIndex == direction_next(entranceDirection)))
+        {
+            mazeEntry |= 0x1 << gMazeCurrentDirectionToHedge[subTileIndex][direction_reverse(nearEntranceTile.direction)];
```

Comment:
```suggestion
            mazeEntry |= 0x1 << gMazeCurrentDirectionToHedge[subTileIndex][entranceDirection];
```
`entranceDirection` is already `direction_reverse(nearEntranceTile.direction)`, is it not?